File: backend-express/src/tests/api/endpoints/auth.test.ts
import supertest from "supertest";

import app from "@/api";
import { closeAll, seedTestData } from "@/lib/setup";
import { getImagePath } from "@/lib/utils";

const request = supertest(app);

beforeAll(async () => {
    await seedTestData();
});

afterAll(async () => {
    await closeAll();
});

describe("POST /api/auth/signup", () => {
    it("responds with json", async () => {
        const response = await request
            .post("/api/auth/signup")
            .field("name", "Didier Drogba")
            .field("email", "new_user@gmail.com")
            .field("password", "very_secret")
            .attach("image", getImagePath("avatar1.jpg"))
            .expect("Content-Type", /json/)
            .expect(200);
        expect(response.body).toHaveProperty("email", "new_user@gmail.com");
        expect(response.body).toHaveProperty("userId");
        expect(response.body).toHaveProperty("access_token");
    });
});

describe("POST /api/auth/signin", () => {
    it("responds with json", async () => {
        const response = await request
            .post("/api/auth/signin")
            .set("Content-Type", "application/x-www-form-urlencoded")
            .send("username=mslimbeji@gmail.com&password=very_secret")
            .expect("Content-Type", /json/)
            .expect(200);
        expect(response.body).toHaveProperty("email", "mslimbeji@gmail.com");
        expect(response.body).toHaveProperty("userId");
        expect(response.body).toHaveProperty("access_token");
    });
});

File: backend-axum/src/tests/api/auth_test.rs
use axum::{
    body::Body,
    http::{Request, StatusCode},
};
use tower::ServiceExt;

use crate::static_;
use crate::tests::api::utils::{MultipartTestRequest, parse_json, setup};

#[tokio::test]
async fn test_signup() {
    let (app, _) = setup().await;

    let (content_type, body_bytes) = MultipartTestRequest::new()
        .add_field("name", "Didier Drogba")
        .add_field("email", "new_user@gmail.com")
        .add_field("password", "very_secret")
        .add_file("image", &static_::get_image_path("avatar1.jpg"))
        .build();

    let response = app
        .oneshot(
            Request::builder()
                .method("POST")
                .uri("/api/auth/signup")
                .header("Content-Type", content_type)
                .body(Body::from(body_bytes))
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(
        response
            .headers()
            .get("Content-Type")
            .expect("test error")
            .to_str()
            .expect("test error")
            .contains("application/json")
    );

    let json = parse_json(response).await;
    assert_eq!(json["email"], "new_user@gmail.com");
    assert!(json.get("user_id").is_some(), "missing user_id");
    assert!(json.get("access_token").is_some(), "missing access_token");
}

#[tokio::test]
async fn test_signin() {
    let (app, _) = setup().await;

    let response = app
        .oneshot(
            Request::builder()
                .method("POST")
                .uri("/api/auth/signin")
                .header("Content-Type", "application/x-www-form-urlencoded")
                .body(Body::from(
                    "username=mslimbeji@gmail.com&password=very_secret",
                ))
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(
        response
            .headers()
            .get("Content-Type")
            .expect("test error")
            .to_str()
            .expect("test error")
            .contains("application/json")
    );

    let json = parse_json(response).await;
    assert_eq!(json["email"], "mslimbeji@gmail.com");
    assert!(json.get("user_id").is_some(), "missing user_id");
    assert!(json.get("access_token").is_some(), "missing access_token");
}
